apps/web/src/hooks/useDocumentParameters/index.ts

```diff
@@ -418,65 +420,70 @@ export function useDocumentParameters<
+  const parameters = useMemo(
+    () => convertToParams(inputsBySource),
+    [inputsBySource],
+  )
+
+  const onMetadataChange = useCallback(
     (metadata: ConversationMetadata) => {
-      setInputs(
-        'manual',
-        recalculateInputs({
-          inputs: inputs.manual.inputs,
-          metadata,
-        }),
-      )
+      setValue((oldState) => {
+        const { doc } = getDocState(oldState, key)
+        console.log('MANUAL_INPUTS', doc.manual.inputs)
+        return updateInputsState({
+          key,
+          source: 'manual',
+          oldState,
+          newInputs: recalculateInputs<'manual'>({
+            inputs: doc.manual.inputs,
+            metadata,
+          }),
+        })
+      })
 
-      // TODO: Remove after a dataset V2 migration
-      // This is not needed with new datasets. We don't store the values
-      // of the cells viewed because they can be modified now.
-      if (datasetVersion === DatasetVersion.V1) {
-        const datasetInputs = recalculateInputs<'dataset'>({
-          inputs: linkedDataset.inputs,
+      setValue((oldState) => {
+        const { doc } = getDocState(oldState, key)
+        const dsId = document.datasetV2Id
+        const datasetDoc = dsId ? doc.datasetV2?.[dsId] : undefined
+        const dsInputs = datasetDoc?.inputs ?? {}
+        const datasetInputs = recalculateInputs<'datasetV2'>({
+          inputs: dsInputs,
           metadata,
         })
-        // Store in local while not stored in the DB
-        setInputs('dataset', datasetInputs)
-
-        if (document.datasetId && linkedDataset) {
-          setDataset({
-            datasetId: document.datasetId,
-            datasetVersion,
-            data: {
-              inputs: datasetInputs,
-              mappedInputs: linkedDataset.mappedInputs,
-              rowIndex: linkedDataset.rowIndex,
-            } as LinkedDataset,
-          })
-        }
-      }
+        return updateInputsState({
+          key,
+          source: 'datasetV2',
+          oldState,
+          newInputs: datasetInputs,
+        })
+      })
 
-      setInputs(
-        'history',
-        recalculateInputs({
-          inputs: inputs.history.inputs,
-          metadata,
-        }),
-      )
+      setValue((oldState) => {
+        const { doc } = getDocState(oldState, key)
+        return updateInputsState({
+          key,
+          source: 'history',
+          oldState,
+          newInputs: recalculateInputs<'history'>({
+            inputs: doc.history.inputs,
+            metadata,
+          }),
+        })
+      })
     },
-    [
-      inputs,
-      setInputs,
-      source,
-      document.datasetId,
-      linkedDataset,
-      datasetVersion,
-      useDatasetsV2,
-    ],
+    [key, document.datasetId, setValue],
   )
 
+  useEffect(() => {
+    if (!metadata) return
+
+    setParameters(Array.from(metadata.parameters))
+    onMetadataChange(metadata)
+  }, [metadata, setParameters])
```

Comment:
This is major change. Before this hook was exporting `onMetadataChange` and this function had a lot of dependencies that was making `useMetadata` being re-rendered and `metadata` was lost. I hate using `useEffect` but I think is better this way.